ert/data/record/_record.py

```diff
@@ -0,0 +1,224 @@
+import pathlib
+from abc import ABC
+from collections import deque
+from enum import Enum
+from typing import Dict, List, Mapping, MutableMapping, Optional, Tuple, Union, cast
+
+from beartype import beartype
+from beartype.roar import BeartypeException  # type: ignore
+from pydantic import PositiveInt
+from ert.serialization import get_serializer
+
+number = Union[int, float]
+numerical_record_data = Union[
+    List[number],
+    Dict[str, number],
+    Dict[int, number],
+]
+blob_record_data = bytes
+record_data = Union[numerical_record_data, blob_record_data]
+record_collection = Tuple["Record", ...]
+RecordIndex = Tuple[Union[int, str], ...]
+
+
+def _build_record_index(
+    data: numerical_record_data,
+) -> RecordIndex:
+    if isinstance(data, MutableMapping):
+        return tuple(data.keys())
+    else:
+        return tuple(range(len(data)))
+
+
+class RecordValidationError(Exception):
+    pass
+
+
+class RecordType(str, Enum):
+    LIST_FLOAT = "LIST_FLOAT"
+    MAPPING_INT_FLOAT = "MAPPING_INT_FLOAT"
+    MAPPING_STR_FLOAT = "MAPPING_STR_FLOAT"
+    BYTES = "BYTES"
+
+
+class Record(ABC):
+    @property
+    def data(self) -> record_data:
+        pass
+
+    @property
+    def record_type(self) -> RecordType:
+        pass
```

Comment:
Any particular reason for not using the `@abstractmethod` decorator?

Reply:
You have a good point as `@abstractmethod` requires the function to be override, which I think is the case here...

Reply:
Thanks for pointing this out. I thought somehow that `ABC` somehow implied this, but I understand now that that was wrong.